File: ice-dss/src/main/java/ice/service/OpenCdsService.java
package ice.service;

import com.sun.xml.ws.client.BindingProviderProperties;
import com.sun.xml.ws.developer.JAXWSProperties;
import ice.dto.support.CdsObjectAssist;
import ice.exception.IceException;
import ice.util.Utilities;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.List;
import java.util.Map;
import javax.xml.datatype.DatatypeConfigurationException;
import javax.xml.datatype.DatatypeFactory;
import javax.xml.ws.BindingProvider;
import org.apache.log4j.Logger;
import org.omg.spec.cdss._201105.dss.DataRequirementItemData;
import org.omg.spec.cdss._201105.dss.EntityIdentifier;
import org.omg.spec.cdss._201105.dss.EvaluationRequest;
import org.omg.spec.cdss._201105.dss.EvaluationResponse;
import org.omg.spec.cdss._201105.dss.FinalKMEvaluationResponse;
import org.omg.spec.cdss._201105.dss.InteractionIdentifier;
import org.omg.spec.cdss._201105.dss.ItemIdentifier;
import org.omg.spec.cdss._201105.dss.KMEvaluationRequest;
import org.omg.spec.cdss._201105.dss.KMEvaluationResultData;
import org.omg.spec.cdss._201105.dss.SemanticPayload;
import org.omg.spec.cdss._201105.dsswsdl.DSSRuntimeExceptionFault;
import org.omg.spec.cdss._201105.dsswsdl.DecisionSupportService;
import org.omg.spec.cdss._201105.dsswsdl.Evaluation;
import org.omg.spec.cdss._201105.dsswsdl.EvaluationExceptionFault;
import org.omg.spec.cdss._201105.dsswsdl.InvalidDriDataFormatExceptionFault;
import org.omg.spec.cdss._201105.dsswsdl.InvalidTimeZoneOffsetExceptionFault;
import org.omg.spec.cdss._201105.dsswsdl.RequiredDataNotProvidedExceptionFault;
import org.omg.spec.cdss._201105.dsswsdl.UnrecognizedLanguageExceptionFault;
import org.omg.spec.cdss._201105.dsswsdl.UnrecognizedScopedEntityExceptionFault;
import org.omg.spec.cdss._201105.dsswsdl.UnsupportedLanguageExceptionFault;
import org.opencds.CdsInput;
import org.opencds.CdsOutput;

/**
 *
 * @author HLN Consulting, LLC
 */
public class OpenCdsService {

    private final static Logger logger = Logger.getLogger(OpenCdsService.class);
    private final static int DEFAULT_TIMEOUT = 10 * 1000;
    private final DecisionSupportService openCdsService = new DecisionSupportService();
    private final Evaluation evaluatePort = openCdsService.getEvaluate();
    private final String endPoint;
    private final int requestTimeout;
    private final int connectTimeout;

    public OpenCdsService(String endPoint) {
        this(endPoint, DEFAULT_TIMEOUT, DEFAULT_TIMEOUT);
    }


    public OpenCdsService(String endPoint, int timeout) {
        this(endPoint, timeout, timeout);
    }

    public OpenCdsService(String endPoint, int requestTimeout, int connectTimeout) {
        super();
        this.endPoint = endPoint;
        this.requestTimeout = requestTimeout;
        this.connectTimeout = connectTimeout;
        Map<String, Object> ctxt = ((BindingProvider) evaluatePort).getRequestContext();
        ctxt.put(JAXWSProperties.HTTP_CLIENT_STREAMING_CHUNK_SIZE, 8192);
        ctxt.put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, endPoint);
        ctxt.put(BindingProviderProperties.REQUEST_TIMEOUT, requestTimeout);
        ctxt.put(BindingProviderProperties.CONNECT_TIMEOUT, connectTimeout);
        ctxt.put("com.sun.xml.internal.ws.connect.timeout", connectTimeout);
        ctxt.put("com.sun.xml.internal.ws.request.timeout", requestTimeout);
        ctxt.put("org.jboss.ws.timeout", requestTimeout);
        ctxt.put(JAXWSProperties.CONNECT_TIMEOUT, connectTimeout);
        ctxt.put(JAXWSProperties.REQUEST_TIMEOUT, requestTimeout);
    }

    public static OpenCdsService getOpenCDS(String endPoint) {
        return new OpenCdsService(endPoint);
    }

    public static OpenCdsService getOpenCDS(String endPoint, int timeout) {
        return new OpenCdsService(endPoint, timeout);
    }

    public static OpenCdsService getOpenCDS(String endPoint, int requestTimeout, int connectTimeout) {
        return new OpenCdsService(endPoint, requestTimeout, connectTimeout);
    }

    public CdsOutput evaluate(CdsInput cdsInput, String scopingEntityId, String businessId, String version, Date executionDate) throws IceException {
        byte[] cdsObjectToByteArray = CdsObjectAssist.cdsObjectToByteArray(cdsInput, CdsInput.class);
        byte[] evaluation = evaluate(cdsObjectToByteArray, scopingEntityId, businessId, version, executionDate);
        CdsOutput cdsOutput = CdsObjectAssist.cdsObjectFromByteArray(evaluation, CdsOutput.class);
        return cdsOutput;
    }

    public byte[] evaluate(byte[] cdsInputByteArray, String scopingEntityId, String businessId, String version, Date executionDate) throws IceException {
        final String METHODNAME = "evaluate ";
        if (logger.isDebugEnabled()) {
            logger.debug(METHODNAME
                    + "calling evaluate with businessId '"
                    + businessId
                    + "' @ "
                    + endPoint
                    + " with requestTimeout:"
                    + requestTimeout
                    + " and connectTimeout:"
                    + connectTimeout);
        }
        long start = System.nanoTime();
        EvaluationResponse response = null;
        byte[] result = null;
        GregorianCalendar gc = new GregorianCalendar();
        gc.setTime(executionDate);

        try {
            InteractionIdentifier interactionIdentifier = getInteractionIdentifier(gc);
            EvaluationRequest evaluationRequest = getEvaluationRequest(cdsInputByteArray, scopingEntityId, businessId, version);
            Utilities.logDuration("evaluate init", start);

            start = System.nanoTime();
            response = evaluatePort.evaluate(interactionIdentifier, evaluationRequest);
            Utilities.logDuration("evaluate execute", start);

            start = System.nanoTime();
            if (response == null) {
                throw new IceException("response is null!");
            }
            List<FinalKMEvaluationResponse> finalKMEvaluationResponse = response.getFinalKMEvaluationResponse();
            if (finalKMEvaluationResponse == null) {
                throw new IceException("finalKMEvaluationResponse is null!");
            }
            if (finalKMEvaluationResponse.size() != 1) {
                throw new IceException("finalKMEvaluationResponse size wrong: " + finalKMEvaluationResponse.size());
            }
            FinalKMEvaluationResponse kmEvaluationResponse = finalKMEvaluationResponse.get(0);
            List<KMEvaluationResultData> kmEvaluationResultData = kmEvaluationResponse.getKmEvaluationResultData();
            if (kmEvaluationResultData == null) {
                throw new IceException("kmEvaluationResultData is null!");
            }
            if (kmEvaluationResultData.size() != 1) {
                throw new IceException("kmEvaluationResultData size wrong: " + kmEvaluationResultData.size());
            }
            KMEvaluationResultData resultData = kmEvaluationResultData.get(0);
            if (resultData == null) {
                throw new IceException("resultData is null!");
            }
            SemanticPayload data = resultData.getData();
            if (data == null) {
                throw new IceException("data is null!");
            }
            List<byte[]> base64EncodedPayload = data.getBase64EncodedPayload();
            if (base64EncodedPayload == null) {
                throw new IceException("base64EncodedPayload is null!");
            }
            if (base64EncodedPayload.size() != 1) {
                throw new IceException("base64EncodedPayload size wrong: " + base64EncodedPayload.size());
            }
            result = base64EncodedPayload.get(0);
            if (result == null) {
                throw new IceException("bytes is null!");
            }

        } catch (DSSRuntimeExceptionFault e) {
            throw new IceException(e.getMessage());
        } catch (EvaluationExceptionFault e) {
            throw new IceException(e.getMessage());
        } catch (InvalidDriDataFormatExceptionFault e) {
            throw new IceException(e.getMessage());
        } catch (InvalidTimeZoneOffsetExceptionFault e) {
            throw new IceException(e.getMessage());
        } catch (RequiredDataNotProvidedExceptionFault e) {
            throw new IceException(e.getMessage());
        } catch (UnrecognizedLanguageExceptionFault e) {
            throw new IceException(e.getMessage());
        } catch (UnrecognizedScopedEntityExceptionFault e) {
            throw new IceException(e.getMessage());
        } catch (UnsupportedLanguageExceptionFault e) {
            throw new IceException(e.getMessage());
        } catch (DatatypeConfigurationException e) {
            throw new IceException(e.getMessage());
        } finally {
            logger.debug(METHODNAME + "end...");
        }
        Utilities.logDuration("evaluate post process", start);
        return result;
    }

    private EntityIdentifier getIIEntityIdentifier(String businessId) {
        EntityIdentifier iiEntityIdentifier = new EntityIdentifier();
        iiEntityIdentifier.setScopingEntityId("gov.nyc.health");
        iiEntityIdentifier.setVersion("1.0.0.0");
        iiEntityIdentifier.setBusinessId(businessId + "Data");
        return iiEntityIdentifier;
    }

    private ItemIdentifier getItemIdentifier(String businessId) {
        ItemIdentifier itemIdentifier = new ItemIdentifier();
        itemIdentifier.setItemId("cdsPayload");
        itemIdentifier.setContainingEntityId(getIIEntityIdentifier(businessId));
        return itemIdentifier;
    }

    private EntityIdentifier getSPEntityIdentifier() {
        EntityIdentifier spEntityIdentifier = new EntityIdentifier();
        spEntityIdentifier.setBusinessId("VMR");
        spEntityIdentifier.setScopingEntityId("org.opencds.vmr");
        spEntityIdentifier.setVersion("1.0");
        return spEntityIdentifier;
    }

    private SemanticPayload getSemanticPayload(byte[] cdsInputByteArray) {
        SemanticPayload semanticPayload = new SemanticPayload();
        semanticPayload.setInformationModelSSId(getSPEntityIdentifier());
        semanticPayload.getBase64EncodedPayload().add(cdsInputByteArray);
        return semanticPayload;
    }

    private DataRequirementItemData getDataRequirementItemData(byte[] cdsInputByteArray, String businessId) {
        DataRequirementItemData dataRequirementItemData = new DataRequirementItemData();
        dataRequirementItemData.setDriId(getItemIdentifier(businessId));
        dataRequirementItemData.setData(getSemanticPayload(cdsInputByteArray));
        return dataRequirementItemData;
    }

    private InteractionIdentifier getInteractionIdentifier(GregorianCalendar gc) throws DatatypeConfigurationException {
        InteractionIdentifier interactionIdentifier = new InteractionIdentifier();
        interactionIdentifier.setInteractionId("123456");
        interactionIdentifier.setScopingEntityId("gov.nyc.health");
        interactionIdentifier.setSubmissionTime(DatatypeFactory.newInstance().newXMLGregorianCalendar(gc));
        return interactionIdentifier;
    }

    private EntityIdentifier getKMEntityIdentifier(String scopingEntityId, String businessId, String version) {
        EntityIdentifier kmEntityIdentifier = new EntityIdentifier();
        kmEntityIdentifier.setScopingEntityId(scopingEntityId);
        kmEntityIdentifier.setVersion(version);
        kmEntityIdentifier.setBusinessId(businessId);
        return kmEntityIdentifier;
    }

    private KMEvaluationRequest getKMEvaluationRequest(String scopingEntityId, String businessId, String version) {
        KMEvaluationRequest kmEvaluationRequest = new KMEvaluationRequest();
        kmEvaluationRequest.setKmId(getKMEntityIdentifier(scopingEntityId, businessId, version));
        return kmEvaluationRequest;
    }

    private EvaluationRequest getEvaluationRequest(byte[] cdsInputByteArray, String scopingEntityId, String businessId, String version) {
        EvaluationRequest evaluationRequest = new EvaluationRequest();
        evaluationRequest.setClientLanguage("");
        evaluationRequest.setClientTimeZoneOffset("");
        evaluationRequest.getKmEvaluationRequest().add(getKMEvaluationRequest(scopingEntityId, businessId, version));
        evaluationRequest.getDataRequirementItemData().add(getDataRequirementItemData(cdsInputByteArray, businessId));
        return evaluationRequest;
    }
}

File: net/OpenCDSNETLib/OpenCDSNETLib/OpenCdsService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ServiceModel;
using ice.service.OpenCds;
using System.Xml.Serialization;
using System.IO;
using System.Xml;

namespace ice.service
{
    public class OpenCdsService
    {
        private static int DEFAULT_TIMEOUT = 10 * 1000;
        private readonly string endPoint;
        private readonly EvaluationClient service;
        private readonly int requestTimeout;
        private readonly int connectTimeout;

        public OpenCdsService(string endPoint)
            : this(endPoint, DEFAULT_TIMEOUT, DEFAULT_TIMEOUT)
        {
        }

        public OpenCdsService(String endPoint, int timeout)
            : this(endPoint, timeout, timeout)
        {
        }

        public OpenCdsService(String endPoint, int requestTimeout, int connectTimeout)
        {
            this.endPoint = endPoint;
            this.requestTimeout = requestTimeout;
            this.connectTimeout = connectTimeout;
            BasicHttpBinding binding = new BasicHttpBinding();
            binding.Name = "OpenCdsService";
            binding.OpenTimeout = TimeSpan.FromMilliseconds(connectTimeout);
            binding.ReceiveTimeout = TimeSpan.FromMilliseconds(requestTimeout);
            binding.MaxBufferSize = 65536;
            binding.MaxBufferPoolSize = 524288;
            binding.ReaderQuotas.MaxDepth = 32;
            binding.ReaderQuotas.MaxStringContentLength = 65536;
            binding.ReaderQuotas.MaxArrayLength = 16384;
            binding.ReaderQuotas.MaxBytesPerRead = 4096;
            binding.ReaderQuotas.MaxNameTableCharCount = 16384;
            EndpointAddress remoteAddress = new EndpointAddress(endPoint);
            service = new EvaluationClient(binding, remoteAddress);
            Console.WriteLine("Constructing OpenCdsService from " + endPoint);
        }
        public CDSOutput evaluate(CDSInput cdsInput, string scopingEntityId, string businessId, string version)
        {
            CDSOutput cdsOutput = null;
            MemoryStream ms = new MemoryStream();
            XmlWriterSettings settings = new XmlWriterSettings();
            settings.Encoding = Encoding.UTF8;
            settings.Indent = true;
            settings.IndentChars = "    ";
            settings.ConformanceLevel = ConformanceLevel.Document;
            XmlWriter writer = XmlTextWriter.Create(ms, settings);
            XmlSerializer serializer = new XmlSerializer(typeof(CDSInput));
            serializer.Serialize(ms, cdsInput);
            string cdsInputString = Encoding.UTF8.GetString(ms.ToArray());
            ms.Close();
            string response = evaluate(cdsInputString, scopingEntityId, businessId, version);
            serializer = new XmlSerializer(typeof(CDSOutput));
            StringReader sr = new StringReader(response);
            cdsOutput = (CDSOutput)serializer.Deserialize(sr);
            sr.Close();
            return cdsOutput;
        }
        public string evaluate(string cdsInputString, string scopingEntityId, string businessId, string version)
        {
            string METHODNAME = "evaluate ";
            if (true)
            {
                Console.WriteLine(METHODNAME
                        + "calling evaluate with businessId '"
                        + businessId
                        + "; scopingEntityId "
                        + scopingEntityId
                        + "; version "
                        + version
                        + "' @ "
                        + endPoint
                        + " with requestTimeout:"
                        + requestTimeout
                        + " and connectTimeout:"
                        + connectTimeout);
            }
            evaluateResponse response = null;
            string result = null;

            InteractionIdentifier interactionIdentifier = getInteractionIdentifier();
            EvaluationRequest evaluationRequest = getEvaluationRequest(encodeTo64(cdsInputString), scopingEntityId, businessId, version);
            evaluate e = new evaluate();

            e.interactionId = interactionIdentifier;
            e.evaluationRequest = evaluationRequest;

            response = service.evaluate(e);

            if (response == null)
            {
                throw new Exception("response is null!");
            }
            EvaluationResponse er = response.evaluationResponse;
            FinalKMEvaluationResponse[] finalKMEvaluationResponse = er.finalKMEvaluationResponse;

            if (finalKMEvaluationResponse == null)
            {
                throw new Exception("finalKMEvaluationResponse is null!");
            }
            if (finalKMEvaluationResponse.Length != 1)
            {
                throw new Exception("finalKMEvaluationResponse size wrong: " + finalKMEvaluationResponse.Length);
            }

            FinalKMEvaluationResponse kmEvaluationResponse = finalKMEvaluationResponse[0];
            KMEvaluationResultData[] kmEvaluationResultData = kmEvaluationResponse.kmEvaluationResultData;

            if (kmEvaluationResultData == null)
            {
                throw new Exception("kmEvaluationResultData is null!");
            }
            if (kmEvaluationResultData.Length != 1)
            {
                throw new Exception("kmEvaluationResultData size wrong: " + kmEvaluationResultData.Length);
            }
            KMEvaluationResultData resultData = kmEvaluationResultData[0];
            if (resultData == null)
            {
                throw new Exception("resultData is null!");
            }

            SemanticPayload data = resultData.data;
            if (data == null)
            {
                throw new Exception("data is null!");
            }
            result = decodeFrom64(data.base64EncodedPayload);
            if (result == null)
            {
                throw new Exception("result is null!");
            }
            return result;
        }

        private EntityIdentifier getIIEntityIdentifier(string businessId)
        {
            EntityIdentifier iiEntityIdentifier = new EntityIdentifier();
            iiEntityIdentifier.scopingEntityId = "gov.nyc.health";
            iiEntityIdentifier.version = "1.0.0.0";
            iiEntityIdentifier.businessId = businessId + "Data";
            return iiEntityIdentifier;
        }

        private ItemIdentifier getItemIdentifier(String businessId)
        {
            ItemIdentifier itemIdentifier = new ItemIdentifier();
            itemIdentifier.itemId = "cdsPayload";
            itemIdentifier.containingEntityId = getIIEntityIdentifier(businessId);
            return itemIdentifier;
        }

        private EntityIdentifier getSPEntityIdentifier()
        {
            EntityIdentifier spEntityIdentifier = new EntityIdentifier();
            spEntityIdentifier.businessId = "VMR";
            spEntityIdentifier.scopingEntityId = "org.opencds.vmr";
            spEntityIdentifier.version = "1.0";
            return spEntityIdentifier;
        }

        private SemanticPayload getSemanticPayload(string cdsInput)
        {
            SemanticPayload semanticPayload = new SemanticPayload();
            semanticPayload.informationModelSSId = getSPEntityIdentifier();
            semanticPayload.base64EncodedPayload = cdsInput;
            return semanticPayload;
        }

        private DataRequirementItemData getDataRequirementItemData(string cdsInput, string businessId)
        {
            DataRequirementItemData dataRequirementItemData = new DataRequirementItemData();
            dataRequirementItemData.driId = getItemIdentifier(businessId);
            dataRequirementItemData.data = getSemanticPayload(cdsInput);
            return dataRequirementItemData;
        }

        private InteractionIdentifier getInteractionIdentifier()
        {
            InteractionIdentifier interactionIdentifier = new InteractionIdentifier();
            interactionIdentifier.interactionId = "123456";
            interactionIdentifier.scopingEntityId = "gov.nyc.health";
            return interactionIdentifier;
        }

        private EntityIdentifier getKMEntityIdentifier(string scopingEntityId, string businessId, string version)
        {
            EntityIdentifier kmEntityIdentifier = new EntityIdentifier();
            kmEntityIdentifier.scopingEntityId = scopingEntityId;
            kmEntityIdentifier.version = version;
            kmEntityIdentifier.businessId = businessId;
            return kmEntityIdentifier;
        }

        private KMEvaluationRequest getKMEvaluationRequest(string scopingEntityId, string businessId, string version)
        {
            KMEvaluationRequest kmEvaluationRequest = new KMEvaluationRequest();
            kmEvaluationRequest.kmId = getKMEntityIdentifier(scopingEntityId, businessId, version);
            return kmEvaluationRequest;
        }

        private EvaluationRequest getEvaluationRequest(string cdsInput, string scopingEntityId, string businessId, string version)
        {
            EvaluationRequest evaluationRequest = new EvaluationRequest();
            evaluationRequest.clientLanguage = "";
            evaluationRequest.clientTimeZoneOffset = "";

            evaluationRequest.kmEvaluationRequest = new KMEvaluationRequest[] { getKMEvaluationRequest(scopingEntityId, businessId, version) };
            evaluationRequest.dataRequirementItemData = new DataRequirementItemData[] { getDataRequirementItemData(cdsInput, businessId) };
            return evaluationRequest;
        }

        static public string encodeTo64(string toEncode)
        {
            byte[] toEncodeAsBytes
                  = System.Text.ASCIIEncoding.ASCII.GetBytes(toEncode);
            string returnValue
                  = System.Convert.ToBase64String(toEncodeAsBytes);
            return returnValue;
        }

        static public string decodeFrom64(string encodedData)
        {
            byte[] encodedDataAsBytes
                  = System.Convert.FromBase64String(encodedData);
            string returnValue =
               System.Text.ASCIIEncoding.ASCII.GetString(encodedDataAsBytes);
            return returnValue;
        }
    }
}
